src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -293,47 +294,23 @@ public int Compare(DocumentId? x, DocumentId? y)
         }
     }
 
-    public async ValueTask<ChecksumsAndIds<DocumentId>> GetChecksumsAndIdsAsync(CancellationToken cancellationToken)
-    {
-        var documentTextChecksums = await SelectAsArrayAsync(
-            static async (state, _, cancellationToken) => await state.GetChecksumAsync(cancellationToken).ConfigureAwait(false),
-            arg: default(VoidResult),
-            cancellationToken).ConfigureAwait(false);
-
-        var documentChecksums = new ChecksumCollection(documentTextChecksums);
-        return new(documentChecksums, SelectAsArray(static s => s.Id));
-    }
-
-    public async ValueTask<ChecksumsAndIds<DocumentId>> GetTextChecksumsAndIdsAsync(CancellationToken cancellationToken)
-    {
-        var documentTextChecksums = await SelectAsArrayAsync(
-            static async (state, _, cancellationToken) =>
-            {
-                var stateChecksums = await state.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-                return stateChecksums.Text;
-            },
-            arg: default(VoidResult),
-            cancellationToken).ConfigureAwait(false);
-
-        var documentChecksums = new ChecksumCollection(documentTextChecksums);
-        return new(documentChecksums, SelectAsArray(static s => s.Id));
-    }
-
     public async ValueTask<DocumentChecksumsAndIds> GetDocumentChecksumsAndIdsAsync(CancellationToken cancellationToken)
     {
-        using var _1 = ArrayBuilder<Checksum>.GetInstance(_map.Count, out var attributeChecksums);
-        using var _2 = ArrayBuilder<Checksum>.GetInstance(_map.Count, out var textChecksums);
+        var attributeChecksums = new Checksum[_map.Count];
+        var textChecksums = new Checksum[_map.Count];
 
+        var index = 0;
         foreach (var (_, state) in _map)
         {
             var stateChecksums = await state.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-            attributeChecksums.Add(stateChecksums.Info);
-            textChecksums.Add(stateChecksums.Text);
+            attributeChecksums[index] = stateChecksums.Info;
+            textChecksums[index] = stateChecksums.Text;
+            index++;
         }
 
         return new(
-            new ChecksumCollection(attributeChecksums.ToImmutableAndClear()),
-            new ChecksumCollection(textChecksums.ToImmutableAndClear()),
+            new ChecksumCollection(ImmutableCollectionsMarshal.AsImmutableArray(attributeChecksums)),
+            new ChecksumCollection(ImmutableCollectionsMarshal.AsImmutableArray(textChecksums)),
             SelectAsArray(static s => s.Id));
```

Comment:
>SelectAsArray [](http://example.com/codeflow?start=12&length=13)

Were the SelectAsArray methods worth changing too?

Reply:
Yup.  Fixed up.